### PR TITLE
Allow mutations to affect cardio

### LINKED ISSUE #52980: Indefatigable and Hyperactive do not raise maximum stamina
### Describe the bug

Indefatigable and Hyperactive do not raise stamina by 25% and 40%, respectively. They instead do not affect maximum stamina whatsoever.

### Steps To Reproduce

1. Using debug mode, record maximum stamina
2. Give yourself Indefatigable and/or Hyperactive
3. Using debug mode, record maximum stamina after receiving aforementioned mutations. Maximum stamina will not have changed.

### Expected behavior

I expected that maximum stamina would have risen after receiving Indefatigable or Hyperactive.

### Screenshots

_No response_

### Versions and configuration

OS: Windows 10
Game version: e5992d2 (tiles)
Graphics version: Tiles
Ingame language: English
Mods loaded: dda, no_npc_food, no_fungal_growth, package_bionic_professions

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

Bugfixes "Allow mutations to affect cardio"

#### Purpose of change

Ever since #51880, the `max_stamina_modifier` mutation attribute (intended as a multiplier on maximum stamina) has been wrongly applied to the cardio fitness calculation (being *added* rather than *multiplied*), making the "Languorous", "Indefatigable" and "Hyperactive" mutations have virtually no effect on maximum stamina.

Fixes #52980 

#### Describe the solution

- [x] Replace the `max_stamina_modifier` mutation attribute with `cardio_multiplier`
- [x] Rebalance multiplier so "Languorous", "Indefatigable", and "Hyperactive" can run approximately the same distances as they could before the cardio system
- [x] Adjust `Character::get_cardiofit()` to apply it as a multiplier to cardio fitness, instead of multiplying stamina
- [x] Document and refactor `Character::update_stamina()`
- [x] Add test cases to cover this and related aspects of cardio

#### Describe alternatives you've considered

The original mutation field `max_stamina_modifier` is presently only used by these three mutations in vanilla, and one mutation in CRIT ("Persistent Body"). Could keep the field around in deprecated form for a while before removing it entirely, but with so few usages, it seemed simpler to just change it.

Keeping the original name of the field would be misleading, since it affects more than just max stamina now, and it was never just a "modifier" but a "multiplier".

#### Testing

Run `tests/cata_test '[cardio]'`

Find some pavement, and run back and forth; count how many steps you can take before getting winded. Try with different mutations.

My measured in-game barefoot running distances on pavement were as follows:

- Before cardio (20aaf467):
  - Normal: 96
  - Languorous: 70
  - Indefatigable: 126
  - Hyperactive: 145
- After cardio, before this PR (9da956ad):
  - Normal: 84
  - Languorous: 81
  - Indefatigable: 88
  - Hyperactive: 90
- After this PR:
  - Normal: 84-85
  - Languorous: 66-67
  - Indefatigable: 103-105
  - Hyperactive: 125-127

#### Additional context

I would like to restore something close to the original balance on these traits, in terms of:

- maximum stamina (relative to a "normal" character)
- stamina regen (likewise)
- how far you can run before getting winded (dependent on the first two)

Maximum stamina used to be exactly 10,000. However, with the new cardio system, maximum stamina depends on cardio fitness level; for a default unmutated character, cardio fitness level is 1738, and maximum stamina is 3500 + 3 * 1738 = 8714.

These mutations used `max_stamina_modifier` to adjust maximum stamina by these amounts:

- Languorous: 0.75 * 8714 = 6535.5
- Indefatigable: 1.25 * 8714 = 10892.5
- Hyperactive: 1.4 * 8714 = 12199.6

So these (approximate) stamina maximums will be the target for the new `cardio_multiplier` equivalent.

Since the most measurable in-game effect of maximum stamina (and regen) is how far you can run before getting winded, I would like to also make that as close as feasible to its former (pre-cardio) behavior.

**Note** See the new test cases for verification on this front.

### CODE REVIEW

**Comment by Photoloss** [comment]
While we're at it why should either version of the "max stamina" modifier intrinsically increase stamina regen? We have a dedicated mutation property for that (not sure whether that one plays nice with cardio) so it would be trivial to remove this hidden effect and instead spell it out explicitly in the mutation data.

**Comment by wapcaplet** [comment]
> While we're at it why should either version of the "max stamina" modifier intrinsically increase stamina regen?

There's detail, discussion, and math at #39994 where it was added. The gist is, we want total stamina regen time to be the same, even if you have more total stamina. So with a higher max, you need more regen to reach the max in the same amount of time.

**Comment by wapcaplet** [comment]
After further testing with various builds (pre-cardio master, post-cardio master, and variations of this branch), I am convinced that the cardio and stamina/regen calculations are definitely broken, or at least wildly skewed from their original behavior with regard to size-altering and cardio-altering mutations.

For instance, HUGE characters can somehow run 10 steps *further* than normal characters in a post-cardio world, and TINY characters run out of breath 10 steps *earlier* than normal characters. This should not be happening.

I am reconsidering the `cardio_multiplier` field now, and trying a more direct transformation using the original `max_stamina_modifier`. I suspect we may want to keep both fields.

**Comment by Terrorforge** [comment]
I don't think they're "broken" so much as this is the inevitably result of tying CARDIO heavily to base metabolic rate. A Huge character has a BMR about twice that of an average-sized character, which results in both a significant immediate increase in stamina, and a long-time increase from CARDIO_ACCUMULATOR due to increased calorie consumption. The reverse would be true of mutations like Tiny.

For the same reason, Fast Metabolism is now a colossal buff to stamina - and mutations like Heat Dependant and Light Eater are significant nerfs.

If you're curious about the effect sizes at play: I've been playing max-height, Fast Metabolism characters that mutate into Ursine or Cattle lately. That puts my final BMR somewhere between 6500 and 7000. Such a character starts with about 12k stamina and immediately jumps to about 24k when they mutate Huge. By my calculations, they would eventually hit 50k+ stamina solely as a result of their natural base metabolism, although it would take something like an entire season to get there.

And aside from the degree of difference perhaps being a bit extreme, I think that mostly makes sense? The tradeoff cold-blooded creatures make is that they're less active, but require fewer resources. The "Tiny" thing seems weird on the surface, but the relevant lines also have mutations like Rapid Metabolism and Hyperactive that, once they're fully functional again, should more than compensate. And idk about the biology if this one, but at least aesthetically it makes sense that enormous bear person could just keep going and going forever.

**Comment by Photoloss** [comment]
Not really, although most of it is just the square cube law which the game happily ignores otherwise. I am not sure what actually determines the short-term endurance we call "stamina" across the entirety of the animal kingdom but there are many small factors at play. Oxygenation, sugar(/ATP?) supply and heat dissipation definitely are important but many animals are built more around their niche lifestyle (grazers, ambush predators, foragers) than what their broader lineage is physically capable of.

Another part of the issue is that we don't really distinguish how exertion scales with character size. The act of dragging a crate of rocks on the ground represents a fixed exertion of force and energy, which means it becomes proportionately easier for larger characters. A similar point could be made for swinging a fixed size sledgehammer although for something lighter like a mace you also need to consider the acceleration of the wielding arm. And running or swimming means you should feel the full impact of square cube scaling because all the energy is deposited within your own body.

And then you add body temperature to the mix where larger bodies benefit from both proportionately lower surface area and higher thermal mass (take longer to cool down) and how much temperature management dictates short-term metabolism in different animals.

TL;DR: it's complicated, we need a Ph. D in biology for this.

**Comment by wapcaplet** [comment]
After spending the vast majority of my time on this PR trying to get the test case in a fairly robust condition, and between that and tedious manual testing (running back and forth and counting steps like a bloody human FitBit), I am happy with how this balanced out in the end, at least for the three mutations I was most concerned about (Languorous, Indefatigable, and Hyperactive). See test results in the PR description.

The test case's `running_turns` function gives a reasonable approximation, but is still pretty far off from the actual in-game running steps I measured. For example, with a baseline normal character, in-game I can run 98 steps (barefoot on pavement) before becoming winded, but the `running_turns` function gives them only 87. Still, I've found it useful for comparison and balancing.

I noted several FIXMEs in the test case where some mutations got a significant buff, possibly too much. For example, despite having lower stamina, all the COLDBLOOD traits can run at least 20 steps further than they could before Cardio, and an "Extreme Metabolism" gives more than 30 extra steps. These tests should make them easier to rebalance, but I'm not doing that part in this PR.

File: tests/cardio_test.cpp [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    // Take a deep breath and start running
+    they.set_stamina( they.get_stamina_max() );
+    they.set_movement_mode( move_mode_run );
+    they.update_body();
+    // Count how many turns it takes to become winded
+    int turns = 0;
+    int STOP = 1000; // Safe exit in case of Superman
+    int last_stamina = they.get_stamina_max();
+    // Run until out of stamina or winded (should happen at the same time)
+    while( they.get_stamina() > 0 && !they.has_effect( effect_winded ) && turns < STOP ) {
```
Comment (kevingranade) [inline tests/cardio_test.cpp:84]: I strongly suspect that the reason this doesn't match up with in-game testing is that one step is not exactly equal to one turn, especially when running.
What you should be doing here instead of counting turns is decrementing u.moves and ticking over to the next turn wwhen moves is less than or equal to zero.

Comment (wapcaplet) [reply tests/cardio_test.cpp]: Thanks, hopefully I grokked right (ef74300). I've ensured that `they.moves` are being decremented by the call to `g->walk_move`, and am now doing `they.moves += they.get_speed()` when it goes to zero or below.

Resulting numbers are a bit *lower* than before (further from in-game observations), so obviously there's something else I'm missing.

Comment (wapcaplet) [reply tests/cardio_test.cpp]: Re-confirming in-game measurements shows they are actually quite close to test results now.

**Comment by BrettDong** [comment]
The numbers seem to be different on current `master` branch: https://github.com/CleverRaven/Cataclysm-DDA/runs/4677585598?check_suite_focus=true

**Comment by BrettDong** [comment]
I tried the new cardio test locally and it passes most of the time, but sometimes (2 out of 30 runs) the actual `running_step()` value fluctuates from what is expected and fails the test.

**Comment by wapcaplet** [comment]
> I tried the new cardio test locally and it passes most of the time, but sometimes (2 out of 30 runs) the actual `running_step()` value fluctuates from what is expected and fails the test.

Ah, darn, I was afraid of this. Each step drains stamina by a randomized amount, so there's a chance of some discrepancy, but that's an awfully big difference - more than could reasonably be handled by a larger `margin`.

In most of my testing, I got away with `margin( 1 )` and changed it to `margin( 3 )` to be generous just before merge, but I guess there's something else I missed.

Thanks for pointing this out, I will investigate.

**Comment by wapcaplet** [comment]
> The numbers seem to be different on current `master` branch: https://github.com/CleverRaven/Cataclysm-DDA/runs/4677585598?check_suite_focus=true

Confirmed running locally with the same `--rng-seed 1641011549`. These are only off by 5-7, so a larger margin could tolerate them better:

```
Given: trait: HUGE
../tests/cardio_test.cpp:132

../tests/cardio_test.cpp:136: FAILED:
  CHECK( running_steps( they ) == Approx( expect_run_tiles ).margin( 3 ) )
with expansion:
  99 == Approx( 106.0 )

Given: trait: GOODCARDIO
../tests/cardio_test.cpp:132

../tests/cardio_test.cpp:136: FAILED:
  CHECK( running_steps( they ) == Approx( expect_run_tiles ).margin( 3 ) )
with expansion:
  98 == Approx( 103.0 )

```
